### PR TITLE
Fixed weird characters in CAN csv

### LINKED ISSUE #54: Empty CAN line produces weird characters
Need to initialize CAN buffer array to empty strings so junk doesn't get written back to the CAN CSV file.

### PR DESCRIPTION
**Pull request by Cam0Cow**
Ensures that no weird characters get inadvertently written to the CAN csv file. Closes #54.

### CODE REVIEW

File: BSP/Simulator/Src/BSP_CAN.c [hunk 1]
```diff
@@ -131,24 +135,18 @@ uint8_t BSP_CAN_Read(CAN_t bus, uint32_t* id, uint8_t* data) {
...
-        }
+    // Rewrite unread entries
+    // Write newline for read entry and any entry that should be there but isn't
+    uint8_t index = 0;
+    for (; index < actualNumCan; index++) {
+        fprintf(fp, "%s", bus == index ? "\n" : currentCAN[index]);
+    }
+    for (; index < NUM_CAN; index++) {
+        fprintf(fp, "\n");
+    }
```
Comment (JimothyGreene) [inline BSP/Simulator/Src/BSP_CAN.c:146]: Did you test that this works properly?

**Review by JimothyGreene — COMMENTED** [review]
It seems to look fine to me. Just a few questions that I have to make sure I'm understanding it all fully.

File: BSP/Simulator/Src/BSP_CAN.c [hunk 2]
```diff
@@ -85,6 +85,7 @@ uint8_t BSP_CAN_Write(CAN_t bus, uint32_t id, uint8_t* data, uint8_t len) {
     }
 
     // Close file
+    fsync(fno);
```
[the remark this answers is not in the document]
Comment (Cam0Cow) [reply BSP/Simulator/Src/BSP_CAN.c:88]: Yes, I was getting inconsistent results when reading immediately after a write

File: BSP/Simulator/Src/BSP_CAN.c [hunk 3]
```diff
@@ -114,9 +115,12 @@ uint8_t BSP_CAN_Read(CAN_t bus, uint32_t* id, uint8_t* data) {
...
     char currentCAN[NUM_CAN][256];
     char csv[256];
-    for (uint8_t i = 0; fgets(csv, 256, fp); i++) {
-        strcpy(currentCAN[i], csv);
+    uint8_t actualNumCan = 0;
+    for (; fgets(csv, 256, fp); actualNumCan++) {
+        strcpy(currentCAN[actualNumCan], csv);
     }
+    // Count how many CAN lines are actually in the csv file
+    if (actualNumCan > NUM_CAN) actualNumCan = NUM_CAN;
```
[the remark this answers is not in the document]
Comment (Cam0Cow) [reply BSP/Simulator/Src/BSP_CAN.c:123]: Not quite, it just makes sure that the number of lines it counted in the file doesn't exceed how many CAN lines we actually have

File: BSP/Simulator/Src/BSP_CAN.c (hunk 1 above)
Comment (Cam0Cow) [reply BSP/Simulator/Src/BSP_CAN.c:146]: Yes, I ran it with an empty file, a file with just one CAN line having data (in either position), and both CAN lines having data. Seemed to work for all of them on my end

**Review by JimothyGreene — CHANGES_REQUESTED** [review]
This looks good in terms of accounting for the proper number of CAN channels. However, when I run the full BSP test with this update, it still runs into the stack smashing issue. I think I found a fix in my branch. This is what I did to the old version of this code: 

```c
    // Re-write entry that wasn't read
    // Leave entry that was read as a blank line
    for (uint8_t i = 0; i < NUM_CAN; i++) {
        if (bus == i) {
            fprintf(fp, "\n");
        } else {
            // Only print data if it wasn't a blank line
            // This setup fixes the bug with stack smashing
            if (strcmp(currentCAN[i], "\n")) {
                fprintf(fp, "%s", currentCAN[i]);
            }
        }
    }
```
So the main thing here is that we only want to write the `currentCAN[i]` string when it's not a newline character. I'm still not 100% why this is the big issue, but doing this fixed the stack smashing. So, I think if you just add that surrounding if block to your code now and don't write a newline in any case besides the above, it should work. Play around with this and the full BSP test to make sure that yours works without the stack smashing.

**Review by JimothyGreene — APPROVED** [review]
Looks good to me

**Pull request merged**